# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsParent.cpp
# Commit: 0e6f9a56563c
# Full Hash: 0e6f9a56563c26cc23ab7a6ef56a3aefae4e1295
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-04-01 21:26:59
# Regressor Bug: 1623278
# File Overlap Count: 1
# Description:
#   Bug 1623278 - Replace GetStorageConnection by HasStorageConnection and MutableStorageConnection accessors. r=dom-workers-and-storage-reviewers,janv
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D67286
# ==============================================================================

diff -r 4c6d84346900 -r 0e6f9a56563c dom/indexedDB/ActorsParent.cpp
--- a/dom/indexedDB/ActorsParent.cpp	Wed Apr 01 09:54:43 2020 +0000
+++ b/dom/indexedDB/ActorsParent.cpp	Wed Apr 01 09:54:56 2020 +0000
@@ -4742,13 +4742,15 @@
 
   NS_INLINE_DECL_THREADSAFE_REFCOUNTING(DatabaseConnection)
 
-  mozIStorageConnection* GetStorageConnection() const {
-    if (mStorageConnection) {
-      AssertIsOnConnectionThread();
-      return mStorageConnection->get();
-    }
-
-    return nullptr;
+  bool HasStorageConnection() const {
+    return static_cast<bool>(mStorageConnection);
+  }
+
+  mozIStorageConnection& MutableStorageConnection() const {
+    AssertIsOnConnectionThread();
+    MOZ_ASSERT(mStorageConnection);
+
+    return **mStorageConnection;
   }
 
   UpdateRefcountFunction* GetUpdateRefcountFunction() const {
@@ -5651,7 +5653,7 @@
 };
 
 class MOZ_STACK_CLASS DatabaseOperationBase::AutoSetProgressHandler final {
-  mozIStorageConnection* mConnection;
+  Maybe<mozIStorageConnection&> mConnection;
 #ifdef DEBUG
   DatabaseOperationBase* mDEBUGDatabaseOp;
 #endif
@@ -5661,7 +5663,7 @@
 
   ~AutoSetProgressHandler();
 
-  nsresult Register(mozIStorageConnection* aConnection,
+  nsresult Register(mozIStorageConnection& aConnection,
                     DatabaseOperationBase* aDatabaseOp);
 
   void Unregister();
@@ -11908,7 +11910,7 @@
   MOZ_ASSERT(connection);
   connection->AssertIsOnConnectionThread();
 
-  MOZ_ASSERT(connection->GetStorageConnection());
+  MOZ_ASSERT(connection->HasStorageConnection());
 
   nsresult rv;
   if (!mUpdateStatement) {
@@ -11942,7 +11944,7 @@
   }
 
   int32_t rows;
-  rv = connection->GetStorageConnection()->GetAffectedRows(&rows);
+  rv = connection->MutableStorageConnection().GetAffectedRows(&rows);
   if (NS_WARN_IF(NS_FAILED(rv))) {
     return rv;
   }
@@ -13834,7 +13836,7 @@
 
   AUTO_PROFILER_LABEL("Database::EnsureConnection", DOM);
 
-  if (!mConnection || !mConnection->GetStorageConnection()) {
+  if (!mConnection || !mConnection->HasStorageConnection()) {
     nsresult rv = gConnectionPool->GetOrCreateConnection(*this, &mConnection);
     if (NS_WARN_IF(NS_FAILED(rv))) {
       return rv;
@@ -20293,7 +20295,7 @@
 }
 
 DatabaseOperationBase::AutoSetProgressHandler::AutoSetProgressHandler()
-    : mConnection(nullptr)
+    : mConnection(Nothing())
 #ifdef DEBUG
       ,
       mDEBUGDatabaseOp(nullptr)
@@ -20311,24 +20313,23 @@
 }
 
 nsresult DatabaseOperationBase::AutoSetProgressHandler::Register(
-    mozIStorageConnection* aConnection, DatabaseOperationBase* aDatabaseOp) {
-  MOZ_ASSERT(!IsOnBackgroundThread());
-  MOZ_ASSERT(aConnection);
+    mozIStorageConnection& aConnection, DatabaseOperationBase* aDatabaseOp) {
+  MOZ_ASSERT(!IsOnBackgroundThread());
   MOZ_ASSERT(aDatabaseOp);
   MOZ_ASSERT(!mConnection);
 
   nsCOMPtr<mozIStorageProgressHandler> oldProgressHandler;
 
   nsresult rv =
-      aConnection->SetProgressHandler(kStorageProgressGranularity, aDatabaseOp,
-                                      getter_AddRefs(oldProgressHandler));
+      aConnection.SetProgressHandler(kStorageProgressGranularity, aDatabaseOp,
+                                     getter_AddRefs(oldProgressHandler));
   if (NS_WARN_IF(NS_FAILED(rv))) {
     return rv;
   }
 
   MOZ_ASSERT(!oldProgressHandler);
 
-  mConnection = aConnection;
+  mConnection = SomeRef(aConnection);
 #ifdef DEBUG
   mDEBUGDatabaseOp = aDatabaseOp;
 #endif
@@ -20345,7 +20346,7 @@
       mConnection->RemoveProgressHandler(getter_AddRefs(oldHandler)));
   MOZ_ASSERT(oldHandler == mDEBUGDatabaseOp);
 
-  mConnection = nullptr;
+  mConnection = Nothing();
 }
 
 MutableFile::MutableFile(nsIFile* aFile, SafeRefPtr<Database> aDatabase,
@@ -21467,7 +21468,7 @@
   }
 
   AutoSetProgressHandler asph;
-  rv = asph.Register(connection, this);
+  rv = asph.Register(*connection, this);
   if (NS_WARN_IF(NS_FAILED(rv))) {
     return rv;
   }
@@ -23119,11 +23120,12 @@
     } else {
       DatabaseConnection* connection = database.GetConnection();
       MOZ_ASSERT(connection);
-      MOZ_ASSERT(connection->GetStorageConnection());
+
+      auto& storageConnection = connection->MutableStorageConnection();
 
       AutoSetProgressHandler autoProgress;
       if (mLoggingSerialNumber) {
-        rv = autoProgress.Register(connection->GetStorageConnection(), this);
+        rv = autoProgress.Register(storageConnection, this);
         if (NS_WARN_IF(NS_FAILED(rv))) {
           SetFailureCode(rv);
         }
@@ -23890,7 +23892,7 @@
   {
     int64_t id;
     MOZ_ALWAYS_SUCCEEDS(
-        aConnection->GetStorageConnection()->GetLastInsertRowID(&id));
+        aConnection->MutableStorageConnection().GetLastInsertRowID(&id));
     MOZ_ASSERT(mMetadata.id() == id);
   }
 #endif
@@ -24064,8 +24066,9 @@
 #ifdef DEBUG
     {
       int32_t deletedRowCount;
-      MOZ_ALWAYS_SUCCEEDS(aConnection->GetStorageConnection()->GetAffectedRows(
-          &deletedRowCount));
+      MOZ_ALWAYS_SUCCEEDS(
+          aConnection->MutableStorageConnection().GetAffectedRows(
+              &deletedRowCount));
       MOZ_ASSERT(deletedRowCount == 1);
     }
 #endif
@@ -24177,9 +24180,7 @@
 
   AUTO_PROFILER_LABEL("CreateIndexOp::InsertDataFromObjectStore", DOM);
 
-  nsCOMPtr<mozIStorageConnection> storageConnection =
-      aConnection->GetStorageConnection();
-  MOZ_ASSERT(storageConnection);
+  auto& storageConnection = aConnection->MutableStorageConnection();
 
   RefPtr<UpdateIndexDataValuesFunction> updateFunction =
       new UpdateIndexDataValuesFunction(this, aConnection);
@@ -24187,14 +24188,14 @@
   NS_NAMED_LITERAL_CSTRING(updateFunctionName, "update_index_data_values");
 
   nsresult rv =
-      storageConnection->CreateFunction(updateFunctionName, 4, updateFunction);
+      storageConnection.CreateFunction(updateFunctionName, 4, updateFunction);
   if (NS_WARN_IF(NS_FAILED(rv))) {
     return rv;
   }
 
   rv = InsertDataFromObjectStoreInternal(aConnection);
 
-  MOZ_ALWAYS_SUCCEEDS(storageConnection->RemoveFunction(updateFunctionName));
+  MOZ_ALWAYS_SUCCEEDS(storageConnection.RemoveFunction(updateFunctionName));
 
   if (NS_WARN_IF(NS_FAILED(rv))) {
     return rv;
@@ -24209,8 +24210,7 @@
   aConnection->AssertIsOnConnectionThread();
   MOZ_ASSERT(mMaybeUniqueIndexTable);
 
-  DebugOnly<void*> storageConnection = aConnection->GetStorageConnection();
-  MOZ_ASSERT(storageConnection);
+  MOZ_ASSERT(aConnection->HasStorageConnection());
 
   // The parameter names are not used, parameters are bound by index only
   // locally in the same function.
@@ -24351,7 +24351,7 @@
   {
     int64_t id;
     MOZ_ALWAYS_SUCCEEDS(
-        aConnection->GetStorageConnection()->GetLastInsertRowID(&id));
+        aConnection->MutableStorageConnection().GetLastInsertRowID(&id));
     MOZ_ASSERT(mMetadata.id() == id);
   }
 #endif
@@ -24966,8 +24966,8 @@
 #ifdef DEBUG
   {
     int32_t deletedRowCount;
-    MOZ_ALWAYS_SUCCEEDS(
-        aConnection->GetStorageConnection()->GetAffectedRows(&deletedRowCount));
+    MOZ_ALWAYS_SUCCEEDS(aConnection->MutableStorageConnection().GetAffectedRows(
+        &deletedRowCount));
     MOZ_ASSERT(deletedRowCount == 1);
   }
 #endif
@@ -25419,7 +25419,7 @@
     DatabaseConnection* aConnection) {
   MOZ_ASSERT(aConnection);
   aConnection->AssertIsOnConnectionThread();
-  MOZ_ASSERT(aConnection->GetStorageConnection());
+  MOZ_ASSERT(aConnection->HasStorageConnection());
 
   AUTO_PROFILER_LABEL("ObjectStoreAddOrPutRequestOp::DoDatabaseWork", DOM);
 
